### PR TITLE
[SPARK-37859][SQL] Do not check for metadata during schema comparison

### PR DESCRIPTION
**Pull request by karenfeng**
### What changes were proposed in this pull request?

Ignores the metadata when comparing the user-provided schema and the actual schema during BaseRelation resolution.

### Why are the changes needed?

Makes it possible to read tables with Spark 3.2 that were written with Spark 3.1, as https://github.com/apache/spark/blob/bd24b4884b804fc85a083f82b864823851d5980c/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L312 added a new metadata field that broke compatibility.

### Does this PR introduce _any_ user-facing change?

Yes. Previously, an error was thrown when a SQL table written with JDBC in Spark 3.1 was read in Spark 3.2. Now, no error is thrown.

### How was this patch tested?

Unit test and manual test with a SQL table written with Spark 3.1.

Query:

```
select * from jdbc_table
```

Before:

```
org.apache.spark.sql.AnalysisException: The user-specified schema doesn't match the actual schema:
```

After: no error

### CODE REVIEW

**Comment by huaxingao** [comment]
@karenfeng Does this need to be in 3.2.1?

**Review by yikf — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -353,7 +353,7 @@ case class DataSource(
       case (dataSource: RelationProvider, Some(schema)) =>
         val baseRelation =
           dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
-        if (baseRelation.schema != schema) {
+        if (baseRelation.schema.withoutMetadata != schema.withoutMetadata) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:356]: seems we can just call `DataType.equalsIgnoreCompatibleNullability(baseRelation.schema, schema)`

**Comment by cloud-fan** [comment]
This seems like a 3.2 regression, shall we include it in 3.2.1?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 1 above)
Comment (yikf) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:356]: It ignore `nullable`, It is possible have a chance to occur failure in query later?  For example, In old 3.1, nullable is true, and there are some null value actually, But nullable is false now, I'm not sure that if have a chance to occur failure of the query.

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:356]: I also think we shouldn't ignore `nullable`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala [hunk 2]
```diff
@@ -508,6 +508,15 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
     f(this) || fields.exists(field => field.dataType.existsRecursively(f))
   }
 
+  private[sql] def withoutMetadata: StructType = {
+    val newFields = fields.map {
+      case StructField(name, dataType, nullable, metadata) =>
```
Comment (huaxingao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala:513]: super nit: `metadata` is not used. Change to `_`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:356]: The user-specified schema is ignored at the end. Technically we can completely ignore it and don't do any checks. The check is to make sure the user-specified schema is not too wrong. If we can ignore the metadata, it seems to me that compatible nullability can also be ignored. e.g. the actual column is not nullable but is marked as nullable in the user-specified schema.

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:356]: Agree. Thanks for the explanation.

**Review by cloud-fan — APPROVED** [review]
LGTM if tests pass

**Comment by cloud-fan** [comment]
`org.apache.spark.sql.sources.TableScanSuite` is failing.

**Comment by cloud-fan** [comment]
The failed test was added to make sure we can't specify the schema: https://github.com/apache/spark/commit/6463e0b9e8067cce70602c5c9006a2546856a9d6#diff-4c2ebb37fe41af1f32693661d48ddee85146a14a6b0803a4955ada357bfd1076R332

Later on, we changed the test to make sure the user-specified schema matches the actual data source schema. I think we should update the test again to use a truly unmatched schema, like `i: STRING`

**Comment by HyukjinKwon** [comment]
offline synced. It's because of a bug in my reverted fix. should be fine to go and merge 👍

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Pull request closed without merging**